### PR TITLE
Distributor: Truncate rather than drop log lines

### LINKED ISSUE #3747: Keep long lines, but truncate them
**Is your feature request related to a problem? Please describe.**
Loki currently discards lines that are too long. Instead of discarding them completely, Loki could simply truncate such lines to the allowed limit, but still keep the line. Loki could also indicate how many bytes were truncated, eg. by adding `...truncated 12345 bytes` at the end of the message. 

Keeping at least first part of the line would allow for searching such lines, despite them not being complete. Furthermore Loki could simplify searching for truncated lines by adding `_truncated_=true` label as well.

### PR DESCRIPTION
**Pull request by MasslessParticle**
- When log lines are longer than maxLineSize, truncate them to maxLineSize rather than dropping them

**Special notes for your reviewer**:
The issue calls out adding a `_truncated_` label but I didn't do that because there doesn't seem to be a way to do it without introducing unexpected behavior: 
- If added before validation, the stream might be dropped for exceeding the max number of label names.
- If added after validation, the number of streams might be doubled for reasons that might surprise an operator. 

This change also doesn't include appending `... truncated 12345 bytes` because it has the potential to introduce strange behavior around `max_line_length`

Fixes #3747

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

**Review by KMiller-Grafana — APPROVED** [review]
Doc addition LGTM.  I have not reviewed the code.

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -300,6 +305,25 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
...
+	if !vContext.maxLineSizeShouldTruncate {
+		return stream.Entries
+	}
+
+	n := 0
+	for _, e := range stream.Entries {
+		if maxSize := vContext.maxLineSize; maxSize != 0 && len(e.Line) > maxSize {
+			truncated := float64(len(e.Line) - maxSize)
+			stream.Entries[n].Line = e.Line[:maxSize]
+			n++
```
Comment (cstyan) [inline pkg/distributor/distributor.go:318]: It looks you're just using n to track which entry in `stream.Entries` to truncate and replace? In that case you can just do `for i, e := range stream.Entries` and where you have `stream.Entries[n].Line`.

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -300,6 +305,25 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
...
 
+func (d *Distributor) truncateLines(vContext validationContext, stream logproto.Stream) []logproto.Entry {
+	if !vContext.maxLineSizeShouldTruncate {
+		return stream.Entries
+	}
+
+	n := 0
+	for _, e := range stream.Entries {
+		if maxSize := vContext.maxLineSize; maxSize != 0 && len(e.Line) > maxSize {
+			truncated := float64(len(e.Line) - maxSize)
```
Comment (dannykopping) [inline pkg/distributor/distributor.go:316]: From your PR comment:
> This change also doesn't include appending ... truncated 12345 bytes because it has the potential to introduce strange behavior around max_line_length

What do you mean by strange behaviour? If we used `...` as the suffix to indicate truncation, could we not just strip an additional 3 characters off the line?

I think truncating a line without any indication is going to cause confusion to users; I think we should try solve this.

This could even be configurable IMHO

File: pkg/validation/validate.go [hunk 3]
```diff
@@ -43,6 +43,16 @@ const (
 	DuplicateLabelNamesErrorMsg = "stream '%s' has duplicate label name: '%s'"
 )
 
+// TruncatedBytes is a metric of the total truncated bytes, by reason.
+var TruncatedBytes = prometheus.NewCounterVec(
```
Comment (dannykopping) [inline pkg/validation/validate.go:47]: Might be worth adding a metric for number of lines truncated too; just knowing the bytes won't give any indication of how often this occurs (or at least will make it harder to determine)

**Review by dannykopping — COMMENTED** [review]
Thanks very much for your awesome contribution @MasslessParticle - we've been meaning to get around to this for a while. Left you a couple comments.

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (MasslessParticle) [reply pkg/distributor/distributor.go:316]: I'm worried about the edge-case where the `max_line_size` is `3` or less. In this case, if a log of length 5 came in and we replaces the last 3 non-truncated characters with `...` we'd end up with a log of length `3`. In that case, we'd would either replace the whole log line or it would still above `max_line_size` and we'd drop it line anyway. In either case, it's unhelpful or confusing what's happening to your logs.

That said, if a line is truncated and then dropped, there are metrics for that and maybe that's the right signal to the operator that the max_line_size is too small?

Another option would be to validate that `max_line_size` is at least 4 which would guarantee that  we could replace the last three bytes with `...` without strange behavior

Comment (MasslessParticle) [reply pkg/distributor/distributor.go:316]: I think the approach I'll take is to just ignore truncation if the `max_line_size` is too small and emit a metric about it.

Comment (dannykopping) [reply pkg/distributor/distributor.go:316]: I don't think we need to emit a metric for that. It's just a corner case (who would set a line size to under 4 bytes?) I think it can just be covered in a unit test.

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (owen-d) [reply pkg/distributor/distributor.go:318]: It looks like n is being used to increment the number of truncated lines, which is then added to a metric.

File: pkg/validation/validate.go (hunk 3 above)
Comment (owen-d) [reply pkg/validation/validate.go:47]: We already have `loki_discarded_{samples,bytes}_total`, which we partition by `user` and a `reason`. Perhaps we should start a `loki_mutated_{samples,bytes}_total` with a `reason=truncated` for consistency (it also leaves open the possibility of adding new things into this model in the future).

**Review by owen-d — COMMENTED** [review]
This is looking great, thanks! A few nits between @dannykopping & myself, but I think we should merge this soon.

File: pkg/validation/validate.go (hunk 3 above)
Comment (dannykopping) [reply pkg/validation/validate.go:47]: Sounds reasonable to me :+1:

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (dannykopping) [reply pkg/distributor/distributor.go:316]: I agree that `...` is less than ideal. If we had a mechanism to add _unindexed labels_, that'd be great. Maybe in the future...
As for the truncation indicator, this isn't something I feel should block the merging of this nice work, but I do feel that we should provide _some_ indication to users that a line was truncated. Maybe a simple flag for the suffix (defaulting to an empty string) could be the way to split the difference here; that way we put the control in users' hands from the get-go.

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (MasslessParticle) [reply pkg/distributor/distributor.go]: This wasn't being used in a metric. I shamelessly copied this pattern without thinking from  the `Push` method to avoid a new slice but didn't catch why it was used there.

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (MasslessParticle) [reply pkg/distributor/distributor.go]: I've pushed a new commit that has the configurable indicator because it seems like a reasonable compromise.

File: pkg/validation/validate.go [hunk 4]
```diff
@@ -43,14 +43,34 @@ const (
 	DuplicateLabelNamesErrorMsg = "stream '%s' has duplicate label name: '%s'"
 )
 
+// TruncatedLines is a metric of the total number of lines mutated, by reason.
+var TruncatedLines = prometheus.NewCounterVec(
```
Comment (dannykopping) [inline pkg/validation/validate.go:47]: `<nit>`Can we rename this to match the metric itself? `TruncatedLines` seems out of sync now with `mutated_samples_total`, and ditto for the other one.`</nit>`

File: pkg/distributor/distributor_test.go [hunk 5]
```diff
@@ -115,6 +115,53 @@ func Test_SortLabelsOnPush(t *testing.T) {
 	require.Equal(t, `{a="b", buzz="f"}`, ingester.pushed[0].Streams[0].Labels)
 }
 
+func Test_TruncateLogLines(t *testing.T) {
```
Comment (dannykopping) [inline pkg/distributor/distributor_test.go:118]: One last thing from my side:
Can you please add a benchmark for this functionality? We should verify that this doesn't greatly increase the ingestion latency.

**Review by dannykopping — COMMENTED** [review]
One more nit and a comment, and then we're good to go from my side.
@owen-d how do you feel about the suffix compromise?

File: pkg/distributor/distributor_test.go (hunk 5 above)
Comment (MasslessParticle) [reply pkg/distributor/distributor_test.go:118]: Benchmarks added and it was a good thing too. Calls to `validation.Mutated{Sample,Bytes}.WithLabelValues` are shockingly expensive. Moving them outside the loop reduced allocs 100x

Comment (MasslessParticle) [reply pkg/distributor/distributor_test.go:118]: Changes that brought allocs down were editing the entry line in place and moving metrics reporting outside of the truncation loop.

Before Changes:
```
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/pkg/distributor
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
Benchmark_SortLabelsOnPush-8                            16393114                61.31 ns/op           16 B/op          1 allocs/op
Benchmark_Push-8                                            1105           1094502 ns/op            1978 B/op         37 allocs/op
Benchmark_PushWithLineTruncation-8                           914           1301114 ns/op            8999 B/op        255 allocs/op
Benchmark_PushWithLineTruncationWithIndicator-8              910           1429702 ns/op           16009 B/op        366 allocs/op
PASS
```

After:
```
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/pkg/distributor
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
Benchmark_SortLabelsOnPush-8                            19185693                61.86 ns/op           16 B/op          1 allocs/op
Benchmark_Push-8                                            1185           1033126 ns/op            1968 B/op         37 allocs/op
Benchmark_PushWithLineTruncation-8                           921           1366202 ns/op            2006 B/op         39 allocs/op
Benchmark_PushWithLineTruncationWithIndicator-8              856           1361395 ns/op            9499 B/op        155 allocs/op
PASS
```

As expected, adding the Indicator bumps allocs a fair bit.

Comment (dannykopping) [reply pkg/distributor/distributor_test.go:118]: Thanks for adding these

I like that the truncation barely adds any alloc overhead. With the indicator, I think it's ok to alloc a bit more as it's user-configurable behaviour and disabled by default. The bytes per op, I'm curious about.

@cyriltovena what do you think?

**Review by owen-d — APPROVED** [review]
LGTM, but I would prefer to remove the `max_line_size_truncate_indicator`. I'm sensitive to making loki _more_ complex when it's already so much so.

File: docs/sources/configuration/_index.md [hunk 6]
```diff
@@ -1781,6 +1781,14 @@ logs in Loki.
 # CLI flag: -distributor.max-line-size
 [max_line_size: <string> | default = none ]
 
+# Truncate log lines when they exceed max_line_size.
+# CLI flag: -distributor.max-line-size-truncate
+[max_line_size_truncate: <boolean> | default = false ]
+
+# When log lines are truncated, append this string.
+# CLI flag: -distributor.max-line-size-truncate-indicator
+[max_line_size_truncate_indicator: <string> | default = "" ]
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply docs/sources/configuration/_index.md:1790]: My only comment here is that indeed we will have metrics to know how many lines were truncated, but customers won't be able to programmatically determine which lines were truncated without an indicator of some sort.

I'm happy to defer to @owen-d's judgment on this one.

File: pkg/distributor/distributor.go [hunk 7]
```diff
@@ -300,6 +305,31 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
...
+	var truncatedSamples, truncatedBytes int
+	for i, e := range stream.Entries {
+		if maxSize := vContext.maxLineSize; maxSize != 0 && len(e.Line) > maxSize {
+			indicator := vContext.maxLineSizeTruncateInd
+			truncateTo := maxSize - len(indicator)
+			if truncateTo <= 0 {
+				continue
+			}
+
+			stream.Entries[i].Line = e.Line[:truncateTo] + indicator
```
Comment (owen-d) [inline pkg/distributor/distributor.go:322]: Not a blocker, but I just realized this can split apart multi-byte runes (https://blog.golang.org/strings)

File: docs/sources/configuration/_index.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (owen-d) [reply docs/sources/configuration/_index.md:1790]: Thanks. @MasslessParticle , please remove the `truncate_indicator` and let's get this merged 🎉 !

[the remark this answers is not in the document]
Comment (MasslessParticle) [reply docs/sources/configuration/_index.md]: Done!

**Review by dannykopping — APPROVED** [review]
LGTM :+1:

**Review by cyriltovena — APPROVED** [review]
LGTM
